# ETL with Pandas

In [1]:
import pandas as pd
import sqlalchemy
import time
import psycopg2

In [2]:
url = "postgresql://postgres:admin@localhost:5433/spark_try" 
engine = sqlalchemy.create_engine(url)
def getRatingPercent(rating):
    return rating/5 * 100
start_time = time.time()

In [3]:
now = time.time()
ratings_df = pd.read_csv("ml-1m/ratings.dat", sep="::", names=['userID', 'movieID', 'rating', 'time'])
print(f"Loading Data rating: {time.time() - now} s")
print(f"Ratings data length: {len(ratings_df)}\n")

<ipython-input-3-4a02e47382a3>:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  ratings_df = pd.read_csv("ml-1m/ratings.dat", sep="::", names=['userID', 'movieID', 'rating', 'time'])


Loading Data rating: 4.295915126800537 s
Ratings data length: 1000209



In [4]:
now = time.time()
movies_df = pd.read_csv("ml-1m/movies.dat", sep="::", names=['movieID', 'movieTitle', 'genres'])
print(f"Loading data film: {time.time() - now} s")
print(f"Film data length: {len(movies_df)}\n")

Loading data film: 0.024960041046142578 s
Film data length: 3883



<ipython-input-4-9a5128c5487f>:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  movies_df = pd.read_csv("ml-1m/movies.dat", sep="::", names=['movieID', 'movieTitle', 'genres'])


In [5]:
now = time.time()
ratings_df['ratingPercent'] = ratings_df['rating'].map(getRatingPercent).astype("int")
print(f"Create rating percentage: {time.time() - now} s\n")

Create rating percentage: 0.22853612899780273 s



In [6]:
now = time.time()
joined_df = pd.merge(ratings_df, movies_df, how="inner", on="movieID")
print(f"Joining ratings and films dataframes: {time.time() - now} s\n")

Joining ratings and films dataframes: 0.1723170280456543 s



In [7]:
now = time.time()
rating_avg_df = joined_df[["movieID", "ratingPercent"]].groupby("movieID").mean()
print(f"Creating average rating for each movieID: {time.time() - now} s\n")

Creating average rating for each movieID: 0.10292601585388184 s



In [9]:
now = time.time()
joined_df.to_sql('film_rating_pandas', engine, index=False)
print(f"Writing dataframe to postgresql: {time.time() - now} s\n")

Writing dataframe to postgresql: 644.5034589767456 s



In [10]:
print(f"Total time: {time.time() - start_time} s")

Total time: 682.631441116333 s


berdasarkan percobaan, waktu yang diperlukan pandas untuk melakukan ETL ke database postgreSQL adalah 
selama 682.631441116333 detik